### PR TITLE
Add customizable adhoc SQL query for select adhoc values

### LINKED ISSUE #330: Allow setup adhoc queries in ConnectionSettings
There are 2 performance issues with how adhoc query gets the values. 

https://github.com/Vertamedia/clickhouse-grafana/blob/d92ff14d529f5e02e835860adf4fa4d58192ed4b/src/adhoc.ts#L3

```
const valuesQuery = "SELECT DISTINCT {field} AS value FROM {database}.{table} LIMIT 300";
```

## issue 1
Happens when cardinality of values is lower than 300. In this case, the engine will scan the entire table to try and get to 300 values. 

I suggest modify to 

```
"SELECT DISTINCT value FROM ( SELECT  {field} AS value FROM {database}.{table}  LIMIT 100000 ) LIMIT 300";
```

The query above limits the total amount of data to scan.

## issue2
The 2nd issue happens with tiered storage setups. The scan is not limited to fast tiered drives. Usually the latest data is located on fast disks so I propose to do

```
"SELECT DISTINCT value FROM ( SELECT  {field} AS value FROM {database}.{table} WHERE  timestamp >= now() - INTERVAL 30 DAY LIMIT 100000 ) LIMIT 300";
```

Although the above case is very specifc to the timestamp field name.

### PR DESCRIPTION
**Pull request by lunaticusgreen**
fix #330 , can be merged after https://github.com/Altinity/clickhouse-grafana/pull/654

### CODE REVIEW

**Comment by alon-cheq** [comment]
Hi [654](https://github.com/Altinity/clickhouse-grafana/pull/654) was merged, any update on this fix?

**Comment by Slach** [comment]
When open or create new Datasource i expected SQL query in adhoc field
image

**Comment by Slach** [comment]
image

**Comment by Slach** [comment]
```
Details
Error: Minified React error #310; visit https://reactjs.org/docs/error-decoder.html?invariant=310 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.

    at QueryEditor (webpack-internal:///./views/QueryEditor/QueryEditor.tsx:81:13)
```

but `QueryEditor.tsx` contains only 78 lines, this is really weird

**Comment by Slach** [comment]
@antip00 instead of create tests for already tested macros, better will create test which cover works with adhoc macros and cover new functionality

**Comment by Slach** [comment]
@lunaticusgreen still get error 
open http://localhost:3000/d/Bnh59Trnk/24-adhoc-2b-24-variable3a-text-formatting?from=now-6h&to=now&var-adhoc_variable=&var-split=service_name&editPanel=2

and remove NL (press <X>)
and after reload adhoc stop works


image

**Review by Slach — APPROVED** [review]

File: src/datasource/datasource.ts [hunk 1]
```diff
@@ -413,7 +416,7 @@ export
   createQuery(options: any, target: any) {
     const queryModel = new SqlQuery(target, this.templateSrv, options);
     // @ts-ignore
-    const adhocFilters = this.templateSrv?.getAdhocFilters(this.adHocFilter?.datasource?.name || 'clickhouse') || []
+    const adhocFilters = getAdhocFilters(this.adHocFilter?.datasource?.name || 'clickhouse',this.uid )
```
Comment (Slach) [inline src/datasource/datasource.ts:419]: Wrong fix
we don't have default data source name
remove `clickhouse` here

**Review by Slach — CHANGES_REQUESTED** [review]

**Review by Slach — APPROVED** [review]

**Comment by Slach** [comment]
@lunaticusgreen e2e test failed

just add new dashboard -> add new visualization (default clickhouse-direct datasource)

got following error
```
An unexpected error happened
Details
TypeError: Cannot read properties of undefined (reading 'uid')

    at QueryEditor (webpack-internal:///./views/QueryEditor/QueryEditor.tsx:82:13)
```

image

**Comment by antip00** [comment]
If you change adhoc variable in panel, it do not changes in reformatted query
[Screencast from 2024-11-22 15-46-12.webm](https://github.com/user-attachments/assets/9b2b910c-ce36-44dd-98a5-7a810700f342)

**Comment by Slach** [comment]
@lunaticusgreen 
it tried `GF_FEATURE_TOGGLES_PUBLICDASHBOARDSSCENE: false` in environment
and it reproduced
`CLICKHOUSE_PLUGIN_VERSION=3.2.3 docker compose up -d grafana_external_install`

http://localhost:3001/d/Bnh59Trnk/24-adhoc-2b-24-variable3a-text-formatting?editPanel=2

all works fine changing in adhoc control changed in Generated SQL
image


looks like we broke it in 2.3.4
`CLICKHOUSE_PLUGIN_VERSION=3.2.4 docker compose up -d grafana_external_install`

the same URL
http://localhost:3001/d/Bnh59Trnk/24-adhoc-2b-24-variable3a-text-formatting?editPanel=2

image

issue-330 branch little bit different behavior

```
git checkout issue-330
git pull origin issue-330
npm run dev
npm run server
```

panel in 3000 port
http://localhost:3000/d/Bnh59Trnk/24-adhoc-2b-24-variable3a-text-formatting?editPanel=2

image


`GRAFANA_VERSION=10.4.3 docker compose up -d grafana`
http://localhost:3000/d/Bnh59Trnk/24-adhoc-2b-24-variable3a-text-formatting?editPanel=2

image

this is definitely bug and shall be fixed

**Pull request merged**